tools/parsecov/main.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+
+	"gopkg.in/urfave/cli.v1"
+)
+
+var (
+	packagePrefixFlagName = "packagePrefix"
+)
+
+type count struct {
+	totalStatements   int
+	coveredStatements int
+}
+
+func (c count) percentage() float32 {
+	return 100 * float32(c.coveredStatements) / float32(c.totalStatements)
+}
+
+func parse(c *cli.Context) error {
+	packageCoverage := make(map[string]count)
+	profile, err := os.Open(c.Args()[0])
+	if err != nil {
+		return err
+	}
+	defer profile.Close()
+
+	// First line is "mode: foo", where foo is "set", "count", or "atomic".
+	// Rest of file is in the format
+	//	encoding/base64/base64.go:34.44,37.40 3 1
+	// where the fields are: name.go:line.column,line.column numberOfStatements count
+
+	// This program only handles set mode
+	s := bufio.NewScanner(profile)
+	s.Scan()
+	line := s.Text()
+	if line != "mode: set" {
+		return fmt.Errorf("invalid coverage mode, expecting 'mode: set' as the first line, but got %q", line)
+	}
+	linenum := 1
+	for s.Scan() {
+		line := s.Text()
+		linenum++
+		lastColon := strings.LastIndex(line, ":")
+		if lastColon == -1 {
+			return fmt.Errorf("line %d invalid: %q", linenum, line)
+		}
+
+		packageName := path.Dir(line[:lastColon])
+		numStatements, covered, err := getStats(line[lastColon+1:])
+		if err != nil {
+			return fmt.Errorf("line %d invalid: %q", linenum, line)
+		}
+		c := packageCoverage[packageName]
+		c.totalStatements += numStatements
+		if covered {
+			c.coveredStatements += numStatements
+		}
+		packageCoverage[packageName] = c
+	}
+
+	packageNames := make([]string, 0, len(packageCoverage))
+	for name := range packageCoverage {
+		packageNames = append(packageNames, name)
+	}
+	sort.Strings(packageNames)
+
+	var totalCount count
+	for _, v := range packageCoverage {
+		totalCount.totalStatements += v.totalStatements
```

Comment:
nit: consider adding a count.Add and count.AddAll ?

Reply:
I think its a bit premature to add an `Add` or `AddAll`, this program is just a script for outputting a summary of coverage, its not a package that anything can depend on. At present it doesn't seem likely to me that it's going to change much in the future so I think extra work to do this may not be useful and also if we did end up changing this, its not clear that Add and Add all would be useful in that new context. So I think just leaving it for now is fine, if we come back to this and want to modify it at that point we will have new information that will help guide any choice of re-structuring.

Reply:
I'm interested though to understand the reasons behind your suggestion for these additional functions?

Reply:
I agree it's not clear, that's why I just mentioned as a nit. I always think it looks a bit fishy when one method has well defined sections of work. Most entangled code starts with a few functions not extracted. But I agree that it's a grey line, I guess I'm often biased towards well defined, small methods, while some people prefer few methods with the expected intentions of the file/type.

feel free to merge